sparse/core.py

```diff
@@ -314,40 +314,54 @@ def __str__(self):
 
     __repr__ = __str__
 
-    def reduction(self, method, axis=None, keepdims=False, dtype=None):
+    @staticmethod
+    def _reduce(method, *args, **kwargs):
+        assert len(args) == 1
+
+        self = args[0]
+        if isinstance(self, scipy.sparse.spmatrix):
+            self = COO.from_scipy_sparse(self)
+
+        return self.reduce(method, **kwargs)
+
+    def reduce(self, method, axis=None, keepdims=False, dtype=None):
+        # Needed for more esoteric reductions like product.
+        self.sum_duplicates()
+
         if axis is None:
             axis = tuple(range(self.ndim))
 
         kwargs = {}
         if dtype:
             kwargs['dtype'] = dtype
 
-        if isinstance(axis, numbers.Integral):
+        if not isinstance(axis, tuple):
             axis = (axis,)
 
         if set(axis) == set(range(self.ndim)):
-            result = getattr(self.data, method)(**kwargs)
+            result = method.reduce(self.data, **kwargs)
         else:
             axis = tuple(axis)
+            neg_axis = tuple(ax for ax in range(self.ndim) if ax not in axis)
+
+            a = self.transpose(neg_axis + axis)
+            a = a.reshape((np.prod([self.shape[d] for d in neg_axis]),
+                           np.prod([self.shape[d] for d in axis])))
+            a.sort_indices()
+
+            flag = np.concatenate(([True], a.coords[0, 1:] != a.coords[0, :-1]))
+            # Partial credit to @shoyer
+            # Ref: https://gist.github.com/shoyer/f538ac78ae904c936844
+            inv_idx, = flag.nonzero()
+            result = method.reduceat(a.data, inv_idx, **kwargs)
+            counts = np.diff(np.concatenate(np.nonzero(flag) + ([a.nnz],)))
+            missing_counts = counts != a.shape[1]
+            result[missing_counts] = method(result[missing_counts],
+                                            _zero_of_dtype(result.dtype))
```

Comment:
This isn't 100% accurate, given that reducing with zeros can potentially change the reduction, but it's a good trade-off and works with most reductions including `max`, `min`, `sum`, `prod`, etc.